README.md
**Version 1.1.0 introduced logic reconsidering based on the [Dart implementation](https://github.com/mitryp/datifyDart). 
That version changed the library configuration approach, allowing to extend the localizations and more.
The configuration after that version is not compatible with the 1.0.X and 0.X.X versions of Datify.
Moreover, most of the methods in the library are deprecated since 1.1.0 and will be removed in 2.0.0, so it will break 
the existing code that uses that methods.**

**Since 1.1.0, the library will warn the users that continue using the deprecated methods.**

## Automatic flexible date extracting from strings in any formats.

**Datify** makes it easy to extract dates from strings in _(nearly)_ any formats.

You will need only to parse the date string with Datify, and it's all good.

The date formats supported by Datify are the following:

* Day first digit-only dates: 20.02.2020, 09/07/2000, 9-1-2005;
* Month first digit-only dates: 02 22 2020, 09.07.2000, 1.9/2005;
* Dates in the general date format: 2020-04-15;
* **Alphanumeric dates in different languages**: 11th of July 2020; 6 липня 2021; 31 декабря, 2021.

See the [Formats](#formats) section for the detailed information about the supported formats.

The behavior of Datify can be configured with the DatifyConfig - see [Configuration](#configuring-datify) section.

### Month name languages supported by default:

- [x] English
- [x] Ukrainian
- [x] Russian

---

## Installing
Simply run `pip install datify` from your command line (pip must be installed).

## Data parsing

To extract a date from a string, use the `.parse(string)` factory of the `Datify` class.
The method takes an input string and the optional parameters `year`, `month`, and `day`.

After that the input string will be parsed. If the optional parameters were given, the respective object fields will
have the provided values.

### Getting the result

After the parsing is done, the result can be retrieved in a different ways:

* If the date is complete, the result can be transformed into a `datetime` object with the `date()` getter.
  > **!** The `date()` getter will be replaced with the `date` property in **2.0.0**.

  However, if the date is incomplete the getter will return None.

  The result is considered complete when the `year`, `month`, and `day` fields of the result are not None.

  To make sure the parsed result is complete and can be transformed to a datetime, the `complete` property is used.
  It returns True if the result is complete and can be transformed int a datetime object.


* To get the not nullable result independent of the parsing result, use the `tuple()` getter.

  It will return a tuple of the following structure: `(day, month, year)` where each element represents the 
  corresponding field of the Datify object.
  > **!** The `tuple()` getter will be replaced with the `tuple` property that returns the tuple of the structure 
    (year, month, day) in **2.0.0**.

* The Datify instance itself has the `year`, `month`, and `day` mutable nullable fields, that can be used to access
  the parsing result.

## Formats

> In the formats below, the sign `$` represents any of the supported date splitters.
>
> The `$?` sign represents an optional separator character (the separator may or may not be present).

- General date format: `YYYY$?MM$?DD` - e.g. _20210706_ or _2022-02-23_ etc;

- `Alphanumeric dates in different languages` - e.g. _6th of July 2021_, _31st of December 2021_, _20 жовтня_, _1 июля_
  etc;
  > Datify tries to find different forms of month names in the natural languages where they are present.

When the `day_first` is set to `true`:

- The most common digit-only date format: `DD$MM$YYYY` - e.g. _20.01.2022_;

When the `day_first` is set to `false`:

- American digit date format (the month is first): `MM$DD$YYYY` - e.g. _12.31.2021_;

> When the `day_first` is set to `false`, Datify will try to find the alphabetic month names before the parsing to avoid
losing the month values in the strings of the format '1 of July 2020'. However, this makes the parsing a bit slower with
this option enabled.

## Configuring Datify

The library behavior can be customized with the `DatifyConfig` class fields and methods.

The following can be customized:

1. Date splitters (`.`, `/`, `-`, ` ` by default).

   Any of the supported splitters can be present in digit-only or alphanumeric dates (See [Formats](#formats) section
   of the documentation).

   To define a new custom separator, it must be added to the `DatifyConfig.splitters` set.

   For instance, to add the `#` separator to the config, the following syntax is used:
   ```python
   DatifyConfig.splitters.add('#')
   ```
   After that the next `Datify.parse()` invocations will use the added splitter in the parsing operations.
   > A splitter can also be string more than one character long


2. Month names localizations, different month aliases.
   By default, Datify supports English, English shortened, Ukrainian and Russian month names:
   `{'january','jan','січень','январь',}`

   More localizations can be added whenever they needed with the `DatifyConfig`:


  * To add a new month name for the specified month, the `DatifyConfig.add_month_name(ordinal: int, name: str)` method
    is used. The `ordinal` argument takes int number in range [1, 12] inclusive to represent the month number.

    For example, to add the French name, `Septembre`, for the 9th month, the following syntax is used:
    ```python
    DatifyConfig.add_month_name(9, 'Septembre')
    ```
    _If the `ordinal` is not in the defined range, the ValueError will be raised._


* To add a new entire localization, which consists of the 12 ordered month names, the
  `DatifyConfig.add_months_locale(locale: Iterable[str])` method is used.

  > The `locale` iterable must have a length of 12 and consist of the unique elements
  If these conditions are not satisfied, the ArgumentError will be thrown.

  For example, to add the French month localization, the following syntax is used:
  ```python
  french_months = (
     'Janvier', 'Février', 'Mars', 'Avril', 'Peut', 'Juin',
     'Juillet', 'Août', 'Septembre', 'Octobre', 'Novembre', 
     'Décembre'
   )
  
  DatifyConfig.add_months_locale(french_months)
  ```
  > Note: The months should be ordered in the months order for the correct work.

> `DatifyConfig` can be accessed with Datify.config field
---

## Example:

Unnecessary code was omitted from the example above. See the `example/datify_example.py` for the full code example.
```python
class Events(abc.ABC):
    """Database emulation for the example.

    This class stores dates and the corresponding event descriptions and provides the method for
    record requesting from the storage.
    """
    _records = {
        Date(year=2021, month=12, day=31): 'New Year party 🎄',
        Date(year=2022, month=1, day=20): 'Birthday celebration 🎁',
        Date(year=2022, month=2, day=14): 'St. Valentines Day 💖',
        Date(year=2022, month=2, day=23): 'The cinema attendance 📽',
        Date(year=2022, month=5, day=23): 'A long-awaited Moment 🔥',
    }
    """Stores the dates and the corresponding event descriptions."""

    @classmethod
    def query(cls, year: int | None = None, month: int | None = None, day: int | None = None) -> str | None:
        """Returns an event descriptions based on the provided date parts.

        If no date parts provided or no corresponding event descriptions are found, the method returns None.
        """

        # handle empty requests
        if all((year is None, month is None, day is None)):
            return None

        # return the first string corresponding to the Date that satisfies the query, if any
        for record_date in cls._records:
            if record_date.satisfies(year, month, day):
                return cls._records[record_date]

        return None


def handle_request(search_request: SearchRequest) -> str:
    """Handles the SearchRequest requests.

    Returns a corresponding event description or the error message.
    """
    date_query = search_request['date']

    # Datify handles all the parsing inside freeing from even thinking about it!
    parsed = Datify.parse(date_query)

    response = Events.query(year=parsed.year, month=parsed.month, day=parsed.day)

    return response if response is not None else 'No events found for this query 👀'


if __name__ == '__main__':
    # define dates in the different formats
    dates = (
        '31.12.2021',  # common digit-only date format
        '2022-02-23',  # another commonly-used date format
        '23-02/2022',  # the supported separators can be combined in the string
        '20 of January',  # date is incomplete but still correctly parsed
        'May',  # just a month name
        '14 лютого 2022',  # Ukrainian date which stands for 14.02.2022
        'not a date',  # not a date at all
    )

    # 'request' all the dates and print the result
    for date in dates:
        print(f'{date}: {handle_request({"date": date})}')
```

The output of the example above:
```plaintext
31.12.2021: New Year party 🎄
2022-02-23: The cinema attendance 📽
23-02/2022: The cinema attendance 📽
20 of January: Birthday celebration 🎁
May: A long-awaited Moment 🔥
14 лютого 2022: St. Valentines Day 💖
not a date: No events found for this query 👀
```


CHANGELOG.md
_The changelog started at ver. 1.1.0_

# 1.1.0

- Major RegExps fix.
- Marked a bunch of methods as deprecated. The 1.1.0 update rebuilds the logic of Datify without breaking the 
existing functionality, but the **2.0.0 version will break the old code**. From this version, the usage of the code 
marked as deprecated will raise DeprecationWarnings. Consider removing the deprecated functionality usage.
- Major reconsidering of the logic of Datify based on the [Dart implementation](https://github.com/mitryp/datifyDart).
  - Completely rewritten the parsing logic.
  - Now the class doesn't raise exceptions when the date is not found.
  - Now the methods are null-safe.
- Replaced the old configuration dictionary with the `DatifyConfig` class **(no backwards compatibility)**. 
See the documentation for more information.
- Added unit tests.
- Updated the README.
- Removed the russian README file.
- Added `example/example.py`.

The list of the deprecated methods:
- is_date_part
- is_date
- find_date
- is_day
- set_day
- is_digit_month
- is_alpha_month
- get_alpha_month
- set_month
- is_year
- set_year
- date_or_tuple
- setup_variables

**These methods will be removed in the 2.0.0 release.** Until then, the usage of these methods will produce 
DeprecationWarnings.

test/datify_test.py
from __future__ import annotations

import unittest
from random import choice, randint

from datify import Datify, DatifyConfig
from datify.datify import _normalize_month_name


class DigitDatesTestCase(unittest.TestCase):
    strings = [
        '31.12.2021',
        '20/01/2022',
        '14 02 2022',
    ]

    def test_days(self):
        for s in self.strings:
            d = Datify.parse(s)
            self.assertEqual(int(s[:2]), d.day)

    def test_digit_months(self):
        for s in self.strings:
            d = Datify.parse(s)
            self.assertEqual(int(s[3:5]), d.month)

    def test_years(self):
        for s in self.strings:
            d = Datify.parse(s)
            self.assertEqual(int(s[6:]), d.year)

    def test_multiline_dates(self):
        date = '''31.
        12
        .2003'''

        d = Datify.parse(date)
        self.assertEqual((31, 12, 2003), d.tuple())


class AlphabeticMonthsTestCase(unittest.TestCase):
    dates = {
        '10 мая 2022': 5,
        '20th of January, 2021': 1,
        '14 лютого 2022': 2,
        '3 of may 2018': 5
    }

    def test_alphabetic_months(self):
        for date in self.dates:
            d = Datify.parse(date)
            self.assertEqual(self.dates[date], d.month)


class GeneralDatesTestCase(unittest.TestCase):
    dates = {
        '20190301': (1, 3, 2019),
        '20220831': (31, 8, 2022),
        '20201201': (1, 12, 2020),
        '2020-01-20': (20, 1, 2020),  # the separators are not supported now
        '2001.12.21': (21, 12, 2001)  # the separators are not supported now
    }

    def test_general_dates(self):
        for date in self.dates:
            d = Datify.parse(date)
            self.assertEqual(self.dates[date], d.tuple())


class MonthFormsTestCase(unittest.TestCase):
    ukrainian = [
        'січня',
        'лютого',
        'березня',
        'квітня',
        'травня',
        'червня',
        'липня',
        'серпня',
        'вересня',
        'жовтня',
        'листопада',
        'грудня',
    ]

    russian = [
        'января',
        'февраля',
        'марта',
        'апреля',
        'мая',
        'июня',
        'июля',
        'августа',
        'сентября',
        'октября',
        'ноября',
        'декабря'
    ]

    def test_month_forms(self):
        def test_month_list(lst: list[int]) -> None:
            for n in range(len(lst)):
                sep = _choose_from_set(Datify.config.splitters)
                date_str = sep.join(map(str,
                                        (randint(1, 31),
                                         lst[n],
                                         randint(2000, 2022))
                                        ))
                self.assertEqual(n + 1, Datify.parse(date_str).month)

        test_month_list(self.ukrainian)
        test_month_list(self.russian)


class IncompleteDatesTestCase(unittest.TestCase):
    dates = {
        '10 of Jan': (10, 1, None),
        'липень 2022': (None, 7, 2022),
        'июнь 2021': (None, 6, 2021),
        '10 2004': (10, None, 2004)
    }

    def test_incomplete_dates(self):
        for date in self.dates:
            d = Datify.parse(date)

            # test the incomplete dates defined correctly
            self.assertEqual(self.dates[date], d.tuple())

            # test the incomplete dates return None on .date() call
            self.assertIsNone(d.date())


class PredefiningDatesTestCase(unittest.TestCase):
    dates = {
        '20200101': (3, 12, 2021),
        '20040120': (31, 12, 2003),
        '11th of June 2004': (11, 7, 2004)
    }

    def test_predefining_dates(self):
        for date in self.dates:
            d = Datify.parse(date, day=self.dates[date][0], month=self.dates[date][1], year=self.dates[date][2])
            self.assertEqual(self.dates[date], d.tuple())


class SettingsTestCase(unittest.TestCase):
    def test_splitters_adding(self):
        sep = '%'
        Datify.config.splitters.add(sep)
        date = 10, 7, 2006

        date_str = sep.join(map(str, date))
        self.assertEqual(date, Datify.parse(date_str).tuple())

        Datify.config.splitters.remove(sep)

    def test_day_first(self):
        dates = [_random_date() for _ in range(100)]
        DatifyConfig.day_first = False

        for date in dates:
            d = Datify.parse(date[0])
            expected = date[1]
            day_month = (expected.day, expected.month)
            expected_tuple = (*(day_month if expected.day > 12 else reversed(day_month)), expected.year)
            self.assertEqual(expected_tuple, d.tuple(),
                             msg=f'{date[0]} should be equal to {expected_tuple}')

        DatifyConfig.day_first = True


class RandomTests(unittest.TestCase):
    tests_count = 10_000

    def test_random(self):
        for i in range(self.tests_count):
            date_str, expected = _random_date(i > self.tests_count / 2)
            d = Datify.parse(date_str)
            self.assertEqual(expected.tuple(), d.tuple(), msg=f'date_str={date_str} was {d.tuple()}')
            self.assertEqual(Datify.parse(date_str).tuple(), d.tuple(), msg='Datify.parse({0}) should be equal to '
                                                                            'Datify({0})'.format(date_str))


class LocalizationTests(unittest.TestCase):
    french_months = (
        'Janvier',
        'Février',
        'Mars',
        'Avril',
        'Peut',
        'Juin',
        'Juillet',
        'Août',
        'Septembre',
        'Octobre',
        'Novembre',
        'Décembre',
    )

    def _cleanup(self):
        for i in range(len(self.french_months)):
            DatifyConfig.months[i].remove(_normalize_month_name(self.french_months[i]))

    def test_locale_addition(self):
        # test that the locale of the wrong length is not added to the config
        self.assertRaises(ValueError, lambda: DatifyConfig.add_months_locale(self.french_months[::2]))

        # test that the correct locale is added correctly
        try:
            DatifyConfig.add_months_locale(self.french_months)
        except ValueError as e:
            self.fail('Unexpected exception raised: {}'.format(e))

        # cleanup
        self._cleanup()

    def test_added_locale_is_processed(self):
        DatifyConfig.add_months_locale(self.french_months)

        dates = {
            '20 septembre 2022': (20, 9, 2022),
            '17 Peut 2020': (17, 5, 2020),
            '2 Avril 2008': (2, 4, 2008),
        }

        for date in dates:
            d = Datify.parse(date)
            self.assertEqual(dates[date], d.tuple())

        # cleanup
        self._cleanup()


def _random_date(is_alphanumeric: bool = False) -> tuple[str, Datify]:
    sep: str = _choose_from_set(DatifyConfig.splitters)
    day = randint(1, 31)
    month_num = randint(1, 12)
    year = randint(1917, 2022)

    return (sep.join(
        map(str, (day, _choose_from_set(DatifyConfig.months[month_num - 1]) if is_alphanumeric else month_num,
                  year))), Datify(year=year, day=day, month=month_num))


def _choose_from_set(st: set[...]) -> ...:
    return choice(tuple(st))


if __name__ == '__main__':
    unittest.main()


datify/deprecation_warning.py
from __future__ import annotations

import warnings
from functools import wraps
from typing import Callable


def deprecated(reason: str, since: str, removed: str | None = None, silent: bool = False):
    """Wrapper used to warn about deprecated functionality in Datify.

    :param reason: the reason of the deprecation
    :param since: the version in which the function was deprecated
    :param removed: the version in which the function is due to being removed
    :param silent: whether to suppress warnings about deprecation or not
    """

    def decorator(func: Callable):
        @wraps(func)
        def wrapper(*args, **kwargs):
            if silent:
                return func(*args, **kwargs)

            warnings.warn(
                f'The `{func.__name__}` is deprecated since {since} ({reason}) and will be removed in '
                + (removed if removed else
                   'future release') + '. Consider not to use it.',
                category=DeprecationWarning, stacklevel=2)
            return func(*args, **kwargs)

        return wrapper

    return decorator


example/datify_example.py
from __future__ import annotations

import abc
import dataclasses

from datify import Datify

SearchRequest = dict[str, str]
"""A representation of a search request for the example.

In the example there should be a 'date' field in the dict to represent the searching by a date.
"""


@dataclasses.dataclass
class Date:
    """The class that represents a date in the example.

    Has nullable year, month, and day fields and a method that checks if the date satisfies the query.
    """

    year: int | None
    month: int | None
    day: int | None

    def __hash__(self):
        return hash((self.year, self.month, self.day))

    def satisfies(self, year: int | None = None, month: int | None = None, day: int | None = None) -> bool:
        """Returns True if the Date object satisfies the given year, month and day.
        If the Date object does not have the field, it considered equal to the given corresponding value.
        The same for the method arguments.
        """

        pairs = ((year, self.year), (month, self.month), (day, self.day))

        return all(
            any((None in pair, pair[0] == pair[1])) for pair in pairs
        )


class Events(abc.ABC):
    """Database emulation for the example.

    This class stores dates and the corresponding event descriptions and provides the method for
    record requesting from the storage.
    """
    _records = {
        Date(year=2021, month=12, day=31): 'New Year party 🎄',
        Date(year=2022, month=1, day=20): 'Birthday celebration 🎁',
        Date(year=2022, month=2, day=14): 'St. Valentines Day 💖',
        Date(year=2022, month=2, day=23): 'The cinema attendance 📽',
        Date(year=2022, month=5, day=23): 'A long-awaited Moment 🔥',
    }
    """Stores the dates and the corresponding event descriptions."""

    @classmethod
    def query(cls, year: int | None = None, month: int | None = None, day: int | None = None) -> str | None:
        """Returns an event descriptions based on the provided date parts.

        If no date parts provided or no corresponding event descriptions are found, the method returns None.
        """

        # handle empty requests
        if all((year is None, month is None, day is None)):
            return None

        # return the first string corresponding to the Date that satisfies the query, if any
        for record_date in cls._records:
            if record_date.satisfies(year, month, day):
                return cls._records[record_date]

        return None


def handle_request(search_request: SearchRequest) -> str:
    """Handles the SearchRequest requests.

    Returns a corresponding event description or the error message.
    """
    date_query = search_request['date']

    # Datify handles all the parsing inside freeing from even thinking about it!
    parsed = Datify.parse(date_query)

    response = Events.query(year=parsed.year, month=parsed.month, day=parsed.day)

    return response if response is not None else 'No events found for this query 👀'


if __name__ == '__main__':
    # define dates in the different formats
    dates = (
        '31.12.2021',  # common digit-only date format
        '2022-02-23',  # another commonly-used date format
        '23-02/2022',  # the supported separators can be combined in the string
        '20 of January',  # date is incomplete but still correctly parsed
        'May',  # just a month name
        '14 лютого 2022',  # Ukrainian date which stands for 14.02.2022
        'not a date',  # not a date at all
    )

    # 'request' all the dates and print the result
    for date in dates:
        print(f'{date}: {handle_request({"date": date})}')
